src/core/stream_recv.c

```diff
@@ -953,10 +953,14 @@ QuicStreamRecvFlush(
         } else if (Status == QUIC_STATUS_PENDING) {
             //
             // The app called the receive complete API inline if
-            // RecvCompletionLength is non-zero.
+            // RecvCompletionInlineCalled is one.
             //
-            FlushRecv = (Stream->RecvCompletionLength != 0);
-
+            if (InterlockedCompareExchange16(
+                (short*)&Stream->RecvCompletionInlineCalled, 0, 1) == 1) {
+                FlushRecv = (Stream->RecvCompletionLength != 0);
+            } else {
+                FlushRecv = FALSE;
+            }
```

Comment:
I don't understand why the old logic wouldn't work. If the app called it inline, then `RecvCompletionLength` is non-zero. Can you please elaborate further to explain why this is necessary?

Reply:
Oh, I see now. I think the better fix is to make the queued receive complete operation that will execute later end up being a no-op, but let me think on this a bit...

Reply:
Or, can we prevent the operation from being queued at all...

Reply:
Perhaps we can do something with the `ReceiveCompleteOperation` pointer.... I need to think on this. But ideally, I think my preference is if the user thread calls receive complete before the callback to the app returns, then I don't want the operation to be queued - msquic will just handle in inline (on its thread) when it returns from the callback. But we have to be careful about the synchronization...

Reply:
I agree with your preference. I will try to modify this pr according to your suggestion.